### PR TITLE
Add getNextViableDate, getPreviousViableDate, getNextViableTime, getPreviousViableTime pre-built functions

### PR DESCRIPTION
**Pull request by cosmicshuai**
closes: #4043 
Assume today is 2020-06-12 and current time is 15:42:21
getNextViableDate:
description: given a timex property of a valid month and day of month, evaluate the next viable date based on the current time at certain timezone
The signature is getNextViableDate(timex: string, timezone? : string) -> result:string
getNextViableDate("XXXX-12-20", "America/Los_Angelas") -> "2020-12-20"
getNextViableDate("XXXX-02-29") -> "2024-02-29"

getPreviousViableDate:
description: given a timex property of a valid month and day of month, evaluate the previous viable date based on the current time at certain timezone
The signature is getPreviousViableDate(timex: string, timezone? : string) -> result:string
getPreviousViableDate("XXXX-12-20", "Eastern Standard Time") -> "2019-12-20"
getPreviousViableDate("XXXX-02-29") -> "2020-02-29"

getNextViableTime:
description: given a timex property of a valid minutes and seconds, evaluate the next viable time based on the current time at certain timezone
The signature is getNextViableTime(timex: string, timezone? : string) -> result:string
getNextViableTime("TXX:12:14", "Asia/Tokyo") -> "T16:12:14"
getNextViableTime("TXX:52:14) -> "T15:52:14"

getPreviousViableTime:
description: given a timex property of a valid minutes and seconds, evaluate the previous viable time based on the current time at certain timezone
The signature is getPreviousViableTime(timex: string, timezone? : string) -> result:string
getPreviousViableTime("TXX:12:14", 'Europe/London') -> "T15:12:14"
getPreviousViableTime("TXX:52:14") -> "T14:52:14"

### CODE REVIEW

**Comment by cosmicshuai** [comment]
@vishwacsena 
Due to there is few references on these functions. I am not quite sure about the signature of getNextViableTime and getPreviousViableTime. Do you have any idea or documentation on these functions? I will change the input and output correspondingly.

**Comment by Danieladu** [comment]
The only concern is how do we ensure that the time base of the comparison is the user’s timeZone, not the timeZone of the machine the sdk is running on.

File: libraries/AdaptiveExpressions/ExpressionFunctions.cs [hunk 1]
```diff
@@ -4137,6 +4137,209 @@ private static IEnumerable<object> Flatten(IEnumerable<object> list, int dept)
...
+                        if (error == null)
+                        {
+                            (parsed, error) = ParseTimexProperty(args[0]);
+                        }
+
+                        if (error == null)
+                        {
+                            if (parsed.Year != null || parsed.Month == null || parsed.DayOfMonth == null)
+                            {
+                                error = $"The paramter {args[0]} must be a timex expression only contains month and day-of-month, for example: XXXX-10-31.";
```
Comment (cleemullins) [inline libraries/AdaptiveExpressions/ExpressionFunctions.cs:4216]: typo on "parameter". 

The grammer on the error message is problematic. Please revise.

Comment (cosmicshuai) [reply libraries/AdaptiveExpressions/ExpressionFunctions.cs:4216]: changed

**Comment by boydc2014** [comment]
> The only concern is how do we ensure that the time base of the comparison is the user’s timeZone, not the timeZone of the machine the sdk is running on.

i think in API layer, at least we can support pass in timezone as a parameter, then later, we can figure out how to get user timezone E2E

File: libraries/AdaptiveExpressions/ExpressionFunctions.cs [hunk 2]
```diff
@@ -4137,6 +4144,277 @@ private static IEnumerable<object> Flatten(IEnumerable<object> list, int dept)
...
+                                validMonth = parsed.Month ?? 0;
+                                validDay = parsed.DayOfMonth ?? 0;
+
+                                if (validMonth == 2 && validDay == 29)
+                                {
+                                    while (!DateTime.IsLeapYear(validYear))
+                                    {
+                                        validYear += 1;
+                                    }
+                                }
```
Comment (chrimc62) [inline libraries/AdaptiveExpressions/ExpressionFunctions.cs:4210]: Nice corner case...

File: tests/AdaptiveExpressions.Tests/ExpressionParserTests.cs [hunk 3]
```diff
@@ -694,7 +694,7 @@ public class ExpressionParserTests
             // init dateTime: 2018-03-15T13:00:00Z
             Test("isDefinite('helloworld')", false),
             Test("isDefinite('2012-12-21')", true),
-            Test("isDefinite('xxxx-12-21')", false),
+            Test("isDefinite('XXXX-12-21')", false),
```
Comment (chrimc62) [inline tests/AdaptiveExpressions.Tests/ExpressionParserTests.cs:697]: >XXXX [](start = 30, length = 4)

Should this actually be case sensitive?

**Comment by chrimc62** [comment]
I do not like these functions.  They really need an at least an optional base datetime to use--which can default to however we express the users timezone/currenttime.  If I say I want to fly "july 2, 2021" and return "july 3rd", then resolving july 3rd should really be wrt the first expression.

File: tests/AdaptiveExpressions.Tests/ExpressionParserTests.cs [hunk 4]
```diff
@@ -789,7 +789,24 @@ public class ExpressionParserTests
             Test("ticks(timestampObj3)", 636503904000000000),
             Test("ticksToDays(2193385800000000)", 2538.64097222),
             Test("ticksToHours(2193385800000000)", 60927.383333333331),
-            Test("ticksToMinutes(2193385811100000)", 3655643.0185), 
+            Test("ticksToMinutes(2193385811100000)", 3655643.0185),
+            Test("isMatch(getPreviousViableDate('XXXX-07-10'), '20[0-9]{2}-07-10')", true),
```
Comment (chrimc62) [inline tests/AdaptiveExpressions.Tests/ExpressionParserTests.cs:793]: >'20[0-9]{2}-07-10' [](start = 63, length = 18)

These tests are not very precise.  If you don't pass in a base date time, then you can calculate wrt to the current customer datetime/timezone.  You can also pass in the base date time which makes these a precise test.

File: tests/AdaptiveExpressions.Tests/ExpressionParserTests.cs (hunk 3 above)
Comment (cosmicshuai) [reply tests/AdaptiveExpressions.Tests/ExpressionParserTests.cs:697]: Yes, I tested it.  new TimexProperty('xxxx-12-21') will return a empty TimexProperty object.

**Comment by cosmicshuai** [comment]
> I do not like these functions. They really need an at least an optional base datetime to use--which can default to however we express the users timezone/currenttime. If I say I want to fly "july 2, 2021" and return "july 3rd", then resolving july 3rd should really be wrt the first expression.

I added a optional parameter for passing a timezone to these functions. If no timezone provided, the default is local timezone of the machine.
However, GitHub was down so that the commit is not showing correctly. I've pushed the commit again.

File: tests/AdaptiveExpressions.Tests/ExpressionParserTests.cs (hunk 4 above)
Comment (cosmicshuai) [reply tests/AdaptiveExpressions.Tests/ExpressionParserTests.cs:793]: add test cases for timezone

Comment (cosmicshuai) [reply tests/AdaptiveExpressions.Tests/ExpressionParserTests.cs:793]: > added test cases for timezone

**Comment by cleemullins** [comment]
Could you please add more detail to either the issue, or the PR description, around what this is and why. The 4 of us doing the PR review now don't have enough data to understand if these functions are correct/sensible or should be changed.

**Comment by cosmicshuai** [comment]
> Could you please add more detail to either the issue, or the PR description, around what this is and why. The 4 of us doing the PR review now don't have enough data to understand if these functions are correct/sensible or should be changed.

Sorry for the missing of the context of these functions. I updated the PR description. 
These functions was dicussed in an email at the start of R10, we will make sure either the issue, or the PR description will provide enough context in the future.

**Comment by cosmicshuai** [comment]
I closed this PR. And rework all these stuffs in another PR.
